### PR TITLE
add save master info interface

### PR DESCRIPTION
**Pull request by anoty**
add save master info interface

### CODE REVIEW

File: canal/dump.go [hunk 1]
```diff
@@ -126,6 +126,8 @@ func (c *Canal) tryDump() error {
 	log.Infof("dump MySQL and parse OK, use %0.2f seconds, start binlog replication at (%s, %d)",
 		time.Now().Sub(start).Seconds(), h.name, h.pos)
 
-	c.master.Update(mysql.Position{h.name, uint32(h.pos)})
+	pos = mysql.Position{h.name, uint32(h.pos)}
+	c.master.Update(pos)
+	c.eventHandler.SaveMasterInfo(pos, false)
```
Comment (siddontang) [inline canal/dump.go]: I think we should use true here.

File: canal/handler.go [hunk 2]
```diff
@@ -11,6 +11,7 @@ type EventHandler interface {
 	OnRow(e *RowsEvent) error
 	OnXID(nextPos mysql.Position) error
 	OnGTID(gtid mysql.GTIDSet) error
+	SaveMasterInfo(pos mysql.Position, force bool) error // Use your own way to save master info
```
Comment (siddontang) [inline canal/handler.go]: Now Canel doesn't know master info now, we should use another name.

Comment (siddontang) [reply canal/handler.go]: Maybe OnPosSynced

**Comment by anoty** [comment]
Done, @siddontang

File: canal/handler.go [hunk 3]
```diff
@@ -11,7 +11,7 @@ type EventHandler interface {
 	OnRow(e *RowsEvent) error
 	OnXID(nextPos mysql.Position) error
 	OnGTID(gtid mysql.GTIDSet) error
-	SaveMasterInfo(pos mysql.Position, force bool) error // Use your own way to save master info
+	OnPosSynced(pos mysql.Position, force bool) error // Use your own way to save master info
```
Comment (siddontang) [inline canal/handler.go]: remove the latter comment. 

Btw, we should add a comment for the force argument.

**Comment by anoty** [comment]
@siddontang done

File: canal/handler.go [hunk 4]
```diff
@@ -11,6 +11,7 @@ type EventHandler interface {
 	OnRow(e *RowsEvent) error
 	OnXID(nextPos mysql.Position) error
 	OnGTID(gtid mysql.GTIDSet) error
+	OnPosSynced(pos mysql.Position, force bool) error // Use your own way to sync position,When force is true,Immediate sync pos.
```
Comment (siddontang) [inline canal/handler.go]: can you move the comment above the function?

Btw, we should add blank after the comma, like
```
Use your own way to sync position. When force is true, sync position immediately.
```

File: cmd/go-canal/main.go [hunk 5]
```diff
@@ -109,6 +109,18 @@ func (h *handler) OnRow(e *canal.RowsEvent) error {
 	return nil
 }
 
+// Note：In asynchronous way hold positions as far as possible
+// because when the data volume is larger
+// delay will cause a master-slave synchronous way
+func (h *handler) OnPosSynced(pos mysql.Position, force bool) error {
```
Comment (siddontang) [inline cmd/go-canal/main.go]: I think we can remove the implementation here.

File: canal/sync.go [hunk 6]
```diff
@@ -119,7 +124,10 @@ func (c *Canal) runSyncBinlog() error {
 			continue
 		}
 
-		c.master.Update(pos)
+		if savePos {
+			c.master.Update(pos)
+			c.eventHandler.OnPosSynced(pos, false)
```
Comment (siddontang) [inline canal/sync.go]: if we need ddl or rotate, maybe we should sync with force?

**Comment by anoty** [comment]
Done @siddontang

**Comment by siddontang** [comment]
LGTM

**Pull request merged**